crates/hir_ty/src/checker.rs

```diff
@@ -84,7 +84,7 @@ impl<'a> FunctionTypeChecker<'a> {
 
     fn check_stmt(&mut self, stmt: &'a hir::Stmt) {
         match stmt {
-            hir::Stmt::ExprStmt { expr, .. } => self.check_expr(*expr),
+            hir::Stmt::Expr { expr, .. } => self.check_expr(*expr),
```

Comment:
**Optimization Issue**: The method 'check_expr' is used to validate each expression recursively. If the expressions are deeply nested, this could lead to performance degradation due to deep recursion. <br> **Fix**: Refactor 'check_expr' to use an iterative approach rather than recursion. This can be achieved by using a loop and a stack to manually manage the expression nodes, which would help in avoiding deep recursion and potentially improve performance. <br> **Code Suggestion**: 
 ```
 - self.check_expr(*expr);
 + self.iterative_check_expr(*expr);
 ```